Events and Agents have a many-to-many relationship

Hwesta opened the pull request:
Previously, in createMETS2, Agents were added to an amdSec by adding everything in the Agents table, then finding all linkingAgentIdentifiers, assume they're User FKs, and creating Agents from that.   Event's linkingAgentIdentifier were added by checking for a linking agent, and if it existed, assume it's a User and create an agent Element, then add all Agents.

With external Agents, not all Agents are associated with every Event, and not all files (amdSecs) have Events associated with every Agent.

This updates the data model so Events and Agents have a many-to-many relationship (instead of only an implicit relationship), and Users have an associated Agent (this is done by adding a UserProfile, which includes a link to the related Agent. Other additional User information could go here too.).

Changes:
- SQL can be run before Django's syncdb/migrate is
  - This was needed to fix column types to what Django expected so Django could create foreign keys for new tables (UserProfile)
- New UserProfile model that links a User to an Agent
- Update insertIntoEvents to take a list of Agents (instead of a single User.id)
- Update digiprovMD creation to handle new Event/Agent relationship
- Create/update an Agent when a User is created/modified
- Update activeAgent (which tracks which user took an action) to refer to the Agent not the User

This still needs:
- [x] <s>Update ansible role to do pre-sync SQL</s>
- [x] Data migrations for things in `Events.linkingAgentIdentifier`
- [x] Create UserProfile for existing users on upgrade

Hwesta committed:
Models: Events and Agents are many-to-many

Hwesta committed:
DatabaseFunctions: insertIntoEvents takes a list of Agents

Update for Events and Agents being many-to-many

Hwesta committed:
METS: Update for Event/Agents many-to-many relationship

Events only link to Agents explicitly associated with them. Agents are only
added to an amdSec if they're associated with Events in that amdSec.

createDigivProvMD creates both Events and Agents, as described. Created a new
createAgent that mimics createEvent in functionality. Update callers.

Add & update tests, update fixtures to be more modular.

Hwesta committed:
Models: Add UserProfile to link to Agent

Add a UserProfile extension to User for additional data, in this case linking
to an associated Agent. Fix primary key field types to what Django expects before
syncdb is run so Django can create the m2m table.

Hwesta committed:
Models: Added post-save signal for User->Agent creation

When a User is saved, create or update an Agent for that user, along with the
UserProfile. Add logging for models & Agent string representation.

Hwesta committed:
MCPServer: Update activeAgent to Agent ID not User ID

When a choice is made by a user, store the activeAgent associated with that
unit as the Agent ID, instead of the user's ID. Update logs to make unitvar
changes clearer.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -443,24 +432,30 @@ def createDigiprovMD(fileUUID):
     """
     Create digiprovMD for PREMIS Events and linking Agents.
     """
+    global globalDigiprovMDCounter
```
:sob:

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -479,60 +474,25 @@ def createEvent(digiprovMD, event_record):
         etree.SubElement(linkingAgentIdentifier, ns.premisBNS + "linkingAgentIdentifierType").text = agent.identifiertype
         etree.SubElement(linkingAgentIdentifier, ns.premisBNS + "linkingAgentIdentifierValue").text = agent.identifiervalue
 
+def createAgent(digiprovMD, agent_record):
```
Maybe keep these as separate parameters so we can call `values` above and avoid the need to instantiate an `Agent` object? We're not using the stuff that a model object would give us.

Hwesta replied:
Possibly - I was modelling this on createEvent above which accepts a record.  Is instantiating an Agent that expensive?

mistydemeo commented on src/MCPServer/lib/linkTaskManagerChoice.py:
```diff
@@ -138,9 +138,12 @@ def xmlify(self):
-            self.unit.setVariable("activeAgent", agent, None)
+    def proceedWithChoice(self, chain, user_id, delayTimerStart=False):
+        if user_id is not None:
+            agent_id = UserProfile.objects.get(user_id=int(user_id)).agent_id
```
Maybe avoid instantiating an `Agent` object here too?

mistydemeo replied:
If we're doing this same thing in a few files, maybe make a helper function to return the `agent_id` as a string?

Hwesta replied:
I don't think it is instantiating an Agent object, just a UserProfile one.  agent_id fetches the value in the column from UserProfile.

I'll look at moving this to utils.

mistydemeo commented:
> Data migrations for things in Events.linkingAgentIdentifier

Would this handle creating the new `many-to-many` relationship for things created before the `create_user_agent` post-save signal existed?

Hwesta commented:
> > Data migrations for things in Events.linkingAgentIdentifier
> 
> Would this handle creating the new many-to-many relationship for things created before the create_user_agent post-save signal existed?

It would have to. This is part of why we're waiting - I'm hoping that Django migrations (including data migrations) will come in first.

Hwesta committed:
Migrations: Events/Agents data migration

Merge migrations adding UserProfile & Agent fields. Add data migration
to convert from old to new Events/Agents relationship.

jrwdunham commented:
This looks good to me.
